compiler/src/dotty/tools/backend/jvm/BTypesFromSymbols.scala

```diff
@@ -197,42 +298,42 @@ class BTypesFromSymbols[I <: BackendInterface](val int: I) extends BTypes {
   final def javaFlags(sym: Symbol): Int = {
 
 
-    val privateFlag = sym.getsJavaPrivateFlag
+    val privateFlag = sym.is(Flags.Private) || (sym.isPrimaryConstructor && sym.owner.isTopLevelModuleClass)
 
-    val finalFlag = sym.getsJavaFinalFlag
+    val finalFlag = sym.is(Flags.Final) &&  !toDenot(sym).isClassConstructor && !(sym.is(Flags.Mutable)) &&  !(sym.enclosingClass.is(Flags.Trait))
 
     import asm.Opcodes._
     GenBCodeOps.mkFlags(
       if (privateFlag) ACC_PRIVATE else ACC_PUBLIC,
-      if (sym.isDeferred || sym.hasAbstractFlag) ACC_ABSTRACT else 0,
+      if (sym.is(Flags.Deferred) || sym.isOneOf(Flags.AbstractOrTrait)) ACC_ABSTRACT else 0,
       if (sym.isInterface) ACC_INTERFACE else 0,
 
       if (finalFlag &&
         // Primitives are "abstract final" to prohibit instantiation
         // without having to provide any implementations, but that is an
         // illegal combination of modifiers at the bytecode level so
         // suppress final if abstract if present.
-        !sym.hasAbstractFlag &&
+        !sym.isOneOf(Flags.AbstractOrTrait) &&
         //  Mixin forwarders are bridges and can be final, but final bridges confuse some frameworks
-        !sym.isBridge)
+        !sym.is(Flags.Bridge))
         ACC_FINAL else 0,
       if (sym.isStaticMember) ACC_STATIC else 0,
-      if (sym.isBridge) ACC_BRIDGE | ACC_SYNTHETIC else 0,
-      if (sym.isArtifact) ACC_SYNTHETIC else 0,
+      if (sym.is(Flags.Bridge)) ACC_BRIDGE | ACC_SYNTHETIC else 0,
+      if (sym.is(Flags.Artifact)) ACC_SYNTHETIC else 0,
       if (sym.isClass && !sym.isInterface) ACC_SUPER else 0,
-      if (sym.hasEnumFlag) ACC_ENUM else 0,
-      if (sym.isVarargsMethod) ACC_VARARGS else 0,
-      if (sym.isSynchronized) ACC_SYNCHRONIZED else 0,
-      if (sym.isDeprecated) asm.Opcodes.ACC_DEPRECATED else 0,
-      if (sym.isEnum) asm.Opcodes.ACC_ENUM else 0
+      if (sym.isAllOf(Flags.JavaEnumTrait)) ACC_ENUM else 0,
+      if (sym.is(Flags.JavaVarargs)) ACC_VARARGS else 0,
+      if (sym.is(Flags.Synchronized)) ACC_SYNCHRONIZED else 0,
+      if (false /*sym.isDeprecated*/) asm.Opcodes.ACC_DEPRECATED else 0,
```

Comment:
Add a TODO that we should add an isDeprecated method in SymUtils.

Reply:
Done